### PR TITLE
[wptrunner] Only load relevant tests

### PR DESCRIPTION
**Pull request by jonathan-j-lee**
Currently, wptrunner selects tests by loading all tests for enabled types via `Manifest.itertypes(...)`, then applying path exclusion/chunking. This is wasteful when only a few tests will run, and is especially slow when running many small subsuites, as each subsuite will reiterate `itertypes(...)`.

This change fixes this by:
1. Resolving a full set of URLs to include in `TestFilter`, instead of adding `--{include,exclude}` to the `IncludeManifest` tree. It's possible to do this because `--{include,exclude}` always supercede any contents of `--include-manifest`.
2. Changing `itertypes(...)` to return another `Manifest`, which is also `Iterable[Tuple[Text, Text, Set[ManifestItem]]]` (i.e., backwards-compatible when used in regular loops/containers).
3. Using `Manifest`'s identity-based hash to cache a map of URL &rightarrow; item (i.e., the manifest is only iterated once).
4. Look up items by include URL directly in `TestFilter.__call__(...)`.

Additional notes:
* The main cost is additional memory that scales with the size of the test suite (e.g., `TestFilter.include_urls` can hold every test ID).
* In Chromium CI on Linux, where many subsuites are run, this reduces the pre-run overhead [from ~50s][1] [to ~10s][2]. This should also improve the local run UX (on my machine, eliminates ~2s when running a single test).
* The existing inclusion/exclusion behavior is captured by new tests to show that the optimization has no functional change.

See also: https://crbug.com/1489723

[1]: https://chromium-swarm.appspot.com/task?id=656255bd09688410
[2]: https://chromium-swarm.appspot.com/task?id=655df0372c36f710

### CODE REVIEW

**Comment by WeizhongX** [comment]
We already plan to overload TestLoader, is it possible in the overloaded TestLoader we load tests directly from tests we want to run?

**Comment by jonathan-j-lee** [comment]
> is it possible in the overloaded TestLoader we load tests directly from tests we want to run?

I don't think that addresses the main constraint, which is that we start with a list of URLs to include, but `manifest.Manifest` is path-keyed because the URLs are not known a priori.

We could try to get tests via `URL -> path -> ManifestItem` or `URL -> ManifestItem`, but building a URL-keyed map in the first place (either here or in `blinkpy`'s `WPTManifest`) has the same asymptotic complexity as `itertypes(...)`. This change relies on the manifest's lazy loading behavior to avoid always incurring that penalty.

**Comment by WeizhongX** [comment]
> addresses

I think we have too kinds of overhead here:
1) When building wpt manifest, it iterates over all tests, even though those won't be run on the bot.
2) At Wptrunner side, iterate over all tests that are not relevant.

I think this CL is trying to address 2?

**Comment by jonathan-j-lee** [comment]
Right, this only addresses (2), which creates per-subsuite overhead and therefore was not noticeable before subsuites landed. (1) is less urgent because its overhead is fixed.

The `manifest/*` changes are only for speeding up directory accesses, not manifest generation.

**Comment by WeizhongX** [comment]
We basically already get manifests and test urls, what we need is test_path, test_type, wpt_test. I think this is possible?

Will look into that more closely. (Doing it this way will be more straightforward and faster, if possible)

(1) could be unavoidable, because we need to get the test url first before we can do sharding.

**Comment by WeizhongX** [comment]
@jgraham can you help take a look?

File: tools/wptrunner/wptrunner/testloader.py
Comment (WeizhongX) [inline tools/wptrunner/wptrunner/testloader.py:308]: Not sure why this is called *group*_tests_by_url? Also this is test_by_url instead of tests_by_url, right?

File: tools/manifest/manifest.py
Comment (WeizhongX) [inline tools/manifest/manifest.py:122]: Should we give this function a different name now? This now does not return an iterable, but instead a filtered_manifest.

Comment (jonathan-j-lee) [reply tools/manifest/manifest.py:122]: The `Manifest` _is_ an `Iterable[Tuple[str, str, Set[ManifestItem]]]` (see `__iter__` above). This is meant to be backwards-compatible when the caller iterates `itertypes(...)` without explicitly invoking `next(...)` (i.e., almost always).

(The `filtered_manifest` initialization is admittedly kind of ugly, but this can be refactored later.)

**Comment by jonathan-j-lee** [comment]
This is finally ready for re-review; PTAL at the updated description. (Should have marked as draft, sorry.)

This approach now fetches the exact items desired, and should be slightly cleaner (and faster for CI) than the directory-based approach. (Unfortunately, it's slightly slower for the few-test case, as in local runs, since the entire manifest is now iterated exactly once again.)

File: tools/wptrunner/wptrunner/testloader.py
Comment (jonathan-j-lee) [reply tools/wptrunner/wptrunner/testloader.py:308]: I agree `group*` is a misnomer, which suggests some collection as a value. Changed to `map_tests_by_url` (it's not `map_test_by_url` because the map holds multiple tests, even though the value is a single test).

**Review by WeizhongX — APPROVED** [review]
There are few disagreement on minor things, but I am OK to land this for now.

**Review by jgraham — CHANGES_REQUESTED** [review]
I think the fastest way to get this to land would be to split `itertypes` into two methods; one that returns a manifest filtered by type and one that iterates over the given types in the current manifest, and then fix the bugs in `resolve_pattern`. But I also think there are bigger wins to have here that will make the code simpler. For example the `manifestinclude` is only used by servo afaik. We could try removing that all together and port servo to one one the other filtering setups (e.g. disabling directories in the metadata or defining default includes/excludes in the virtual testsuite).

File: tools/wptrunner/wptrunner/manifestinclude.py
Comment (jonathan-j-lee) [reply tools/wptrunner/wptrunner/manifestinclude.py:102]: Apparently not, since `testloader.ManifestLoader.load()` still returns `Dict[manifest.Manifest, Dict[str, str]]`:

https://github.com/web-platform-tests/wpt/blob/aeaebad53eab2288fb9d26ba9f2888361ead74aa/tools/wptrunner/wptrunner/testloader.py#L331-L334

which is plumbed through to `testloader.TestLoader.manifests` and consumed here in `resolve_pattern()`.

(It seems worthwhile to add `TestRoot.url_base` and pass the same `TestRoot` everywhere, but that's for another PR.)

File: tools/wptrunner/wptrunner/testloader.py
Comment (jonathan-j-lee) [reply tools/wptrunner/wptrunner/testloader.py:297]: Ok, renamed to `test_by_url`

File: tools/manifest/manifest.py
Comment (jonathan-j-lee) [reply tools/manifest/manifest.py]: I found a better way: Simply don't load irrelevant test types in the first place. There's already some support for this; this has been extended by plumbing `--test-types -> testloader.ManifestLoader(types=...) -> manifest.load_and_update(types=...)` and dropping irrelevant types after the update. This also unifies the behavior of `load_and_update(types=...)` for the update/no-update cases.

File: tools/wptrunner/wptrunner/manifestinclude.py
Comment (jonathan-j-lee) [reply tools/wptrunner/wptrunner/manifestinclude.py]: > Before this change we'd run all the variants, now we don't run anything.

Not sure what you mean. I think `iterpath()` still returns all variants? See `test_filter_resolve_prefix_to_variants` (added).

(`iterdir()`, despite the name, also does works on file paths, so deduplicated the URLs at the end.)

> It seems likely you want to strip off the url base, rather than the path separator.

Good catch, thanks!

**Comment by jonathan-j-lee** [comment]
> For example the manifestinclude is only used by servo afaik.

Well, recently also Ladybird:

https://github.com/SerenityOS/serenity/blob/169d24ae2ec8b3d6774894ffacd4a96caf59a1aa/Tests/LibWeb/WPT/run.sh#L68

Tackle this another time.

**Comment by jgraham** [comment]
> Tackle this another time.

Yes, sorry I didn't mean to imply that this issue should block on that.

> Well, recently also Ladybird

It looks like they're literally just running one directory, so it would be pretty easy to convert them to a new system :)

File: tools/wptrunner/wptrunner/manifestinclude.py
Comment (jgraham) [reply tools/wptrunner/wptrunner/manifestinclude.py]: `/wpt run firefox /url/url-constructor.any.html` runs 4 tests on master, but 0 after this PR is applied.

**Comment by jonathan-j-lee** [comment]
Not pursuing this anymore; decided to land a patch downstream (https://crrev.com/c/5606147).

I did spin off the tests into #46648 since I had already gone through the trouble of writing them. Take a look?

**Pull request closed without merging**
